lib/SILOptimizer/Mandatory/TFDeabstraction.cpp

```diff
@@ -2552,49 +2559,63 @@ void TFDeabstraction::evaluateAttributesAndDoPacking(
                                      context, &opBuilder);
   }
   // Okay, if we got this far then we have all valid attributes and inputs.
-  // Figure out the result list.
-  SmallVector<Type, 4> resultTypes;
-  auto userResultTy = origResult->getType().getASTType();
-  if (!tf::flattenTensorFlowValueAggregate(userResultTy, resultTypes)) {
-    auto loc = getUserSourceLocation(origInst);
-    diagnose(fn.getModule().getASTContext(), loc.getSourceLoc(),
-             diag::tf_op_result_not_value_or_aggregate, userResultTy)
-      .highlight(loc.getSourceRange());
-    return;
-  }
-  auto resultSILTypes = map<SmallVector<SILType, 8>>(resultTypes, [&](Type ty) {
-      return SILType::getPrimitiveObjectType(ty->getCanonicalType()); });
 
+  // Time to create the op and deal with results.
   auto loc = getUserSourceLocation(origInst);
-  auto op = opBuilder.build(B, context, loc, resultSILTypes);
-
-  // Recursively pack results to a value with the user-specified aggregate type.
-  auto resultIt = op->getResults().begin();
-  std::function<SILValue(SILType)> packResultsToAggregate;
-  packResultsToAggregate = [&](SILType aggregateTy) -> SILValue {
-    if (isTensorFlowValue(aggregateTy))
-      return *resultIt++;
-    if (auto tupleTy = aggregateTy.getAs<TupleType>()) {
-      SmallVector<SILValue, 8> elts;
-      for (auto i : range(tupleTy->getNumElements()))
-        elts.push_back(
-            packResultsToAggregate(aggregateTy.getTupleElementType(i)));
-      return B.createTuple(loc, aggregateTy, elts);
-    }
-    if (auto *decl = aggregateTy.getStructOrBoundGenericStruct()) {
-      SmallVector<SILValue, 8> elts;
-      for (auto *field : decl->getStoredProperties())
-        elts.push_back(packResultsToAggregate(
-            aggregateTy.getFieldType(field, B.getModule())));
-      return B.createStruct(loc, aggregateTy, elts);
+
+  // If there is only one result, it may be an aggregate that we have to pack.
+  if (origInst->getNumResults() == 1) {
```

Comment:
do we also need to cover the case of unpacking an output param (that's an aggregate)?

Reply:
No, the output param conversion is part of this same pass and it always runs before this part and either succeeds or errors out. There's `llvm_unreachable("Attributes cannot be output parameters");` which will trigger if there is still an output param.

I do think that `simplifyOperands` (which handles out params) and `evaluateAttributesAndDoPacking` (this function) should eventually be combined. They both iterate over operands and results and make improvements to them, and I don't think there's a good reason any more to do that in two phases. This will make it clearer that out params are already handled. I'll leave this refactoring for later though.